src/controller/CommissioneeDeviceProxy.cpp

```diff
@@ -0,0 +1,331 @@
+/*
+ *
+ *    Copyright (c) 2020-2021 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *  @file
+ *    This file contains implementation of Device class. The objects of this
+ *    class will be used by Controller applications to interact with CHIP
+ *    devices. The class provides mechanism to construct, send and receive
+ *    messages to and from the corresponding CHIP devices.
+ */
+
+#include <controller/CommissioneeDeviceProxy.h>
+
+#include <controller-clusters/zap-generated/CHIPClusters.h>
+
+#include <app/CommandSender.h>
+#include <app/ReadPrepareParams.h>
+#include <app/util/DataModelHandler.h>
+#include <lib/core/CHIPCore.h>
+#include <lib/core/CHIPEncoding.h>
+#include <lib/core/CHIPSafeCasts.h>
+#include <lib/support/CodeUtils.h>
+#include <lib/support/ErrorStr.h>
+#include <lib/support/SafeInt.h>
+#include <lib/support/logging/CHIPLogging.h>
+
+using namespace chip::Callback;
+
+namespace chip {
+
+// TODO - Refactor LoadSecureSessionParametersIfNeeded() as device object is no longer persisted
+CHIP_ERROR CommissioneeDeviceProxy::LoadSecureSessionParametersIfNeeded(bool & didLoad)
+{
+    didLoad = false;
+
+    // If there is no secure connection to the device, try establishing it
+    if (mState != ConnectionState::SecureConnected)
+    {
+        ReturnErrorOnFailure(LoadSecureSessionParameters());
+        didLoad = true;
+    }
+    else
+    {
+        if (mSecureSession.HasValue())
+        {
+            Transport::SecureSession * secureSession = mSessionManager->GetSecureSession(mSecureSession.Value());
+            // Check if the connection state has the correct transport information
+            if (secureSession->GetPeerAddress().GetTransportType() == Transport::Type::kUndefined)
+            {
+                mState = ConnectionState::NotConnected;
+                ReturnErrorOnFailure(LoadSecureSessionParameters());
+                didLoad = true;
+            }
+        }
+        else
+        {
+            mState = ConnectionState::NotConnected;
+            ReturnErrorOnFailure(LoadSecureSessionParameters());
+            didLoad = true;
+        }
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR CommissioneeDeviceProxy::SendCommands(app::CommandSender * commandObj)
+{
+    bool loadedSecureSession = false;
+    ReturnErrorOnFailure(LoadSecureSessionParametersIfNeeded(loadedSecureSession));
```

Comment:
What happens to commandObj if LoadSecureSessionParametersIfNeeded does in fact need to initiate the async NewPairing?

I am generally suspicious of this auto-connect-if-not-connected logic.  Does this actually work?  Will commandObj actually be sent?

Reply:
The auto-connect part of the logic needs to be reworked/removed. I have added a TODO here to get to it in the follow on cleanup
https://github.com/project-chip/connectedhomeip/pull/10808/files#diff-8f9de4143e87a3aad6e7fce591bf052bd08e83a037737cfd6f57f61c3ee8e3e5R46

The logic was mainly added since we were storing PASE session parameters in persistent storage, and recreating the session from it. It's no longer needed. But, this PR is growing to more than it should be. So splitting that work as a follow on PR, as a smaller change.